src/HttpCacheListener.php

```diff
@@ -273,4 +293,68 @@ public function setVary(Headers $headers)
 
         return $this;
     }
+
+    /**
+     * @param Headers $headers
+     * @param HttpResponse $response
+     * @return $this
+     */
+    public function setEtag(Headers $headers, HttpResponse $response)
+    {
+        if (! empty($this->cacheConfig['etag'])
+            && (! $headers->has('etag')
+                || ! empty($this->cacheConfig['etag']['override']))
+        ) {
+            $etag = new Header\Etag($this->generateEtag($response));
+            $headers->addHeader($etag);
+        }
+
+        return $this;
+    }
+
+    /**
+     * @param HttpRequest $request
+     * @param HttpResponse $response
+     * @return $this
+     */
+    public function setNotModified(HttpRequest $request, HttpResponse $response)
+    {
+        if (!$request->getHeaders()->has('If-Match') || !$response->getHeaders()->has('Etag')) {
+            return $this;
+        }
+
+        $requestEtags = $request->getHeaders()->get('If-Match')->getFieldValue();
+        $requestEtags = !is_array($requestEtags) ? [$requestEtags] : $requestEtags;
+        $responseEtag = $response->getHeaders()->get('Etag')->getFieldValue();
+
+        if (in_array($responseEtag, $requestEtags) || in_array('*', $requestEtags)) {
+            $response->setStatusCode(304);
+            $response->setContent(null);
+        }
+
+        return $this;
+    }
+
+    /**
+     * Generates an Etag based on the response.
+     *
+     * @param HttpResponse $response
+     * @return string Etag
+     */
+    protected function generateEtag(HttpResponse $response)
+    {
+        $generator = new DefaultEtagGenerator();
```

Comment:
I didn't see the `config/module.config.php` before. I'll add the `etag` option there.

Cons of that approach:
 * users always have to provide a generator and a correctly configured container. When using a default generator if no generator is given we can avoid that repetition.
 * if a user doesn't provide a generator, it would silently fail (we could throw an exception instead of returning `void`)
 * the generator can already be easily overwritten in the config
 * etags are only on the header if the module config contains a `etag` section

What do you think?